conductor/lib/uploader.py

```diff
@@ -295,9 +333,79 @@ def do_upload(self, upload_url, filename, md5):
                                                 tries=1,
                                                 use_api_key=True)
 
+    @common.DecRetry(retry_exceptions=api_client.CONNECTION_EXCEPTIONS, tries=5)
+    def do_multipart_upload(self, job, filename, md5):
+        '''
+        Files larger than 5GB will be split into 1GB parts and hydrated in S3
+        once all parts are uploaded.
+
+        On successful part upload to S3, it will return an ETag. This value must be
+        tracked along with the part number in order to complete and hydrate the file in S3.
 
-class Uploader(object):
+        The multipart completion call can take a long time on larger files. Tests on
+        larger files (~100GB) took between 1 and 2 seconds to complete.
+        '''
+        uploads = []
+        complete_payload = {
+            "upload_id": job["upload_id"],
+            "hash": md5,
+            "completed_parts": [],
+            "project": self.project
+        }
+
+        # iterate over parts and upload
+        for part in job["parts"]:
+            resp = self.do_part_upload(part["url"], filename, part_number=part["number"])
+            uploads.append(resp)
+            completed_part = {
+                "PartNumber": part["number"],
+                "ETag": resp.headers['ETag'].strip('"')
+            }
+            complete_payload["completed_parts"].append(completed_part)
+
+        # Complete multipart upload in order to hydrate file in S3 for availability
+        uri_path = '/api/files/v2/complete_multipart'
+        headers = {'Content-Type': 'application/json'}
+        self.api_client.make_request(uri_path=uri_path,
+                                     verb='POST',
+                                     headers=headers,
+                                     data=json.dumps(complete_payload),
+                                     raise_on_error=True,
+                                     use_api_key=True)
+
+        return uploads
+
+    def do_part_upload(self, upload_url, filename, part_number=None):
```

Comment:
Why does `part_number` have a default value? This function will raise an exception if it is ever `None`.